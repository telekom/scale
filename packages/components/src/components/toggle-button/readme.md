# scale-toggle-button



<!-- Auto Generated Below -->


## Properties

| Property                     | Attribute                      | Description                                                                                                                                            | Type                                       | Default        |
| ---------------------------- | ------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------------------ | -------------- |
| `ariaDescriptionTranslation` | `aria-description-translation` | a11y text for getting meaningful value. `$buttonNumber` and `$selected` are template variables and will be replaces by their corresponding properties. | `string`                                   | `'$selected'`  |
| `ariaLabel`                  | `aria-label`                   | (optional) aria-label attribute needed for icon-only buttons                                                                                           | `string`                                   | `undefined`    |
| `ariaLangDeselected`         | `aria-lang-deselected`         | (optional) translation of 'deselected                                                                                                                  | `string`                                   | `'deselected'` |
| `ariaLangSelected`           | `aria-lang-selected`           | (optional) translation of 'selected                                                                                                                    | `string`                                   | `'selected'`   |
| `background`                 | `background`                   | (optional) Button background                                                                                                                           | `"grey" \| "white"`                        | `'white'`      |
| `colorScheme`                | `color-scheme`                 | <span style="color:red">**[DEPRECATED]**</span> - variant should replace colorScheme<br/><br/>                                                         | `"color" \| "monochrome"`                  | `'color'`      |
| `disabled`                   | `disabled`                     | (optional) If `true`, the button is disabled                                                                                                           | `boolean`                                  | `false`        |
| `hideBorder`                 | `hide-border`                  | (optional) set the border-radius left, right or both                                                                                                   | `boolean`                                  | `undefined`    |
| `iconOnly`                   | `icon-only`                    | (optional) Button type                                                                                                                                 | `boolean`                                  | `false`        |
| `iconPosition`               | `icon-position`                | (optional) Icon position related to the label                                                                                                          | `"after" \| "before"`                      | `'before'`     |
| `position`                   | `position`                     | (optional) position within group                                                                                                                       | `number`                                   | `undefined`    |
| `radius`                     | `radius`                       | (optional) set the border-radius left, right or both                                                                                                   | `"both" \| "left" \| "neither" \| "right"` | `null`         |
| `selected`                   | `selected`                     | (optional) If `true`, the button is selected                                                                                                           | `boolean`                                  | `false`        |
| `size`                       | `size`                         | (optional) The size of the button                                                                                                                      | `"large" \| "regular" \| "small" \| "xs"`  | `'regular'`    |
| `styles`                     | `styles`                       | (optional) Injected CSS styles                                                                                                                         | `string`                                   | `undefined`    |
| `toggleButtonId`             | `toggle-button-id`             | (optional) toggle button's id                                                                                                                          | `string`                                   | `undefined`    |
| `variant`                    | `variant`                      | (optional) background variant of a selected toggle-button                                                                                              | `"color" \| "monochrome"`                  | `'color'`      |


## Events

| Event         | Description                                                                                        | Type                                              |
| ------------- | -------------------------------------------------------------------------------------------------- | ------------------------------------------------- |
| `scale-click` | Emitted when button is clicked                                                                     | `CustomEvent<{ id: string; selected: boolean; }>` |
| `scaleClick`  | <span style="color:red">**[DEPRECATED]**</span> in v3 in favor of kebab-case event names<br/><br/> | `CustomEvent<{ id: string; selected: boolean; }>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
