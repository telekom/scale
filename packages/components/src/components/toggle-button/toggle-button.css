/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  --width: auto;
  --spacing-x: var(--scl-spacing-24);
  --spacing-x-icon-only: var(--scl-spacing-8);
  --min-height: var(--scl-spacing-24);
  --height-xs: var(--scl-spacing-24);
  --height-small: var(--scl-spacing-32);
  --height-regular: var(--scl-spacing-40);
  --height-large: var(--scl-spacing-48);
  --radius: var(--scl-radius-4);
  --transition: all var(--scl-motion-duration-fast)
      var(--scl-motion-easing-standard),
    border-radius 0ms;
  --box-shadow-focus: inset 0 0 0 var(--scl-spacing-2) var(--scl-color-focus);
  --font-weight: var(--scl-font-weight-bold);
  --font-size-large: var(--scl-font-size-16);
  --font-size-small: var(--scl-font-size-14);
  --font-size-xs: var(--scl-font-size-12);
  --line-height: var(--scl-font-line-height-150);
  --spacing-icon-x: var(--scl-spacing-8);
  --vertical-align: middle;
  --border-color: var(--scl-color-grey-90);
  --border-color-disabled: var(--scl-color-background-disabled);

  /* size: small */
  --font-size-small: var(--scl-font-size-12);
  --line-height-small: var(--scl-font-line-height-133);
  --min-height-small: var(--scl-spacing-32);

  /* variant: primary */
  --radius-primary: var(--radius);
  --background-primary: var(--scl-color-grey-10);
  --background-primary-disabled: var(--scl-color-grey-0);
  --color-primary: var(--scl-color-grey-90);
  --color-primary-hover: var(--scl-color-primary-hover);
  --color-primary-active: var(--scl-color-primary-active);
  --color-primary-disabled: var(--scl-color-grey-20);

  /* variant: secondary different from primary*/
  --background-secondary: var(--scl-color-white);

  /* selected */
  --color-selected: var(--scl-color-white);
  --background-selected-light: var(--scl-color-primary);
  --background-selected-hover-light: var(--color-primary-hover);
  --background-selected-active-light: var(--scl-color-primary-active);
  --background-selected-dark: var(--scl-color-grey-90);
  --background-selected-hover-dark: var(--scl-color-grey-70);
  --background-selected-active-dark: var(--scl-color-grey-90);

  /* high contrast */
  --color-high-contrast: var(--scl-color-white);
}

.toggle-button {
  box-sizing: border-box;
  display: inline-flex;
  align-items: center;
  position: relative;
  border: 0;
  outline: none;
  cursor: pointer;
  user-select: none;
  font-family: inherit;
  word-spacing: inherit;
  letter-spacing: inherit;
  justify-content: center;
  text-decoration: none;
  font-weight: var(--font-weight);
  font-size: var(--font-size);
  line-height: var(--line-height);
  min-height: var(--min-height);
  width: var(--width);
  padding-left: var(--spacing-x);
  padding-right: var(--spacing-x);
  vertical-align: var(--vertical-align);
  transition: var(--transition);
  margin: 0;
}

.toggle-button--primary {
  background: var(--background-primary);
}

.toggle-button--secondary {
  background: var(--background-secondary);
}

.toggle-button--icon-before ::slotted(*) {
  margin-right: var(--spacing-icon-x);
  pointer-events: none;
}

.toggle-button--icon-after ::slotted(*) {
  margin-left: var(--spacing-icon-x);
  pointer-events: none;
}

.toggle-button--icon-only ::slotted(*) {
  pointer-events: none;
}

.toggle-button--xs {
  height: var(--height-xs);
  font-size: var(--font-size-xs);
}

.toggle-button--small {
  height: var(--height-small);
  font-size: var(--font-size-small);
}

.toggle-button--regular {
  height: var(--height-regular);
  font-size: var(--font-size-large);
}

.toggle-button--large {
  height: var(--height-large);
  font-size: var(--font-size-large);
}

.toggle-button:not(.button--disabled):focus {
  box-shadow: var(--box-shadow-focus);
}

.toggle-button:not(.button--disabled):hover {
  color: var(--color-primary-hover);
}

.toggle-button:not(.button--disabled):active {
  color: var(--scl-color-primary-active);
}

.toggle-button:disabled {
  color: var(--color-primary-disabled);
  pointer-events: none;
  border: 1px solid var(--border-color-disabled);
}

.toggle-button:disabled.toggle-button--primary {
  background: var(--background-primary-disabled);
}

.toggle-button--selected.toggle-button--color {
  color: var(--color-selected);
  background: var(--background-selected-light);
}

.toggle-button--selected:not(.button--disabled).toggle-button--color:hover {
  color: var(--color-selected);
  background: var(--background-selected-hover-light);
}

.toggle-button--selected:not(.button--disabled).toggle-button--color:active {
  color: var(--color-selected);
  background: var(--background-selected-active-light);
}

.toggle-button--selected.toggle-button--monochrome {
  color: var(--color-selected);
  background: var(--background-selected-dark);
}

.toggle-button--selected:not(.button--disabled).toggle-button--monochrome:hover {
  color: var(--color-selected);
  background: var(--background-selected-hover-dark);
}

.toggle-button--selected:not(.button--disabled).toggle-button--monochrome:active {
  color: var(--color-selected);
  background: var(--background-selected-active-dark);
}

.toggle-button--border {
  border: 1px solid var(--border-color);
}

.toggle-button--left,
.toggle-button--left:disabled {
  border-right: 0;
  border-radius: var(--radius-primary) 0 0 var(--radius-primary);
}

.toggle-button--right,
.toggle-button--right:disabled {
  border-left: 0;
  border-radius: 0 var(--radius-primary) var(--radius-primary) 0;
}

.toggle-button--both,
.toggle-button--both:disabled {
  border-radius: var(--radius-primary);
}

.toggle-button--all,
.toggle-button--all:disabled {
  border-radius: var(--radius-primary);
}
.toggle-button--neither,
.toggle-button--neither:disabled {
  border-right: 0;
  border-left: 0;
  border-radius: 0;
}

@media screen and (forced-colors: active), (-ms-high-contrast: active) {
  .toggle-button {
    color: var(--color-high-contrast);
  }
}
