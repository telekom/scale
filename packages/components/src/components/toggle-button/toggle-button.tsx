/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import {
  Component,
  Prop,
  h,
  Host,
  Element,
  Event,
  EventEmitter,
  Method,
} from '@stencil/core';
import classNames from 'classnames';
import { emitEvent } from '../../utils/utils';
import statusNote from '../../utils/status-note';

enum iconSizes {
  xs = '12',
  small = '16',
  regular = '22',
  large = '24',
}

let i = 0;

@Component({
  tag: 'scale-toggle-button',
  styleUrl: 'toggle-button.css',
  shadow: true,
})
export class ToggleButton {
  @Element() hostElement: HTMLElement;
  /** (optional) The size of the button */
  @Prop() size?: 'large' | 'regular' | 'small' | 'xs' = 'regular';
  /** (optional) Button background */
  @Prop() background?: 'grey' | 'white' = 'white';
  /** @deprecated - variant should replace colorScheme */
  @Prop() colorScheme?: 'monochrome' | 'color' = 'color';
  /** (optional) background variant of a selected toggle-button */
  @Prop() variant?: 'monochrome' | 'color' = 'color';
  /** (optional) If `true`, the button is disabled */
  @Prop() disabled?: boolean = false;
  /** (optional) If `true`, the button is selected */
  @Prop({ mutable: true }) selected?: boolean = false;
  /** (optional) Button type */
  @Prop() iconOnly?: boolean = false;
  /** (optional) Icon position related to the label */
  @Prop({ reflect: true, mutable: true }) iconPosition: 'before' | 'after' =
    'before';
  /** (optional) set the border-radius left, right or both */
  @Prop() hideBorder: false;
  /** (optional) set the border-radius left, right or both */
  @Prop() radius: 'left' | 'right' | 'both' | 'neither' | null = null;
  /** (optional) toggle button's id */
  @Prop({ reflect: true }) toggleButtonId?: string;
  /** (optional) aria-label attribute needed for icon-only buttons */
  @Prop() ariaLabelToggleButton: string;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;
  /** (optional) position within group */
  @Prop() position?: number;
  /** (optional) translation of 'selected */
  @Prop() ariaLangSelected? = 'selected';
  /** (optional) translation of 'deselected */
  @Prop() ariaLangDeselected? = 'deselected';
  /** a11y text for getting meaningful value. `$buttonNumber` and `$selected` are template variables and will be replaces by their corresponding properties.  */
  @Prop() ariaDescriptionTranslation = '$selected';
  /** Emitted when button is clicked */
  @Event({ eventName: 'scale-click' }) scaleClick!: EventEmitter<{
    id: string;
    selected: boolean;
  }>;
  /** @deprecated in v3 in favor of kebab-case event names */
  @Event({ eventName: 'scaleClick' }) scaleClickLegacy!: EventEmitter<{
    id: string;
    selected: boolean;
  }>;

  hasScaleIcon = false;

  private focusableElement: HTMLElement;

  @Method()
  async setFocus() {
    this.focusableElement.focus();
  }

  connectedCallback() {
    this.setIconPositionProp();
    this.handleIconShape();
  }

  componentDidLoad() {
    this.handleIconSize();
  }

  componentDidRender() {
    this.handleIconSize();
    if (this.hostElement.hasAttribute('aria-label')) {
      statusNote({
        tag: 'deprecated',
        message:
          'Property "ariaLabel" is deprecated. Please use the "ariaLabelToggleButton" property!',
        type: 'warn',
        source: this.hostElement,
      });
    }
  }

  componentWillLoad() {
    if (this.toggleButtonId == null) {
      this.toggleButtonId = 'toggle-button-' + i++;
    }
  }

  getAriaDescriptionTranslation() {
    const replaceSelected = this.selected
      ? this.ariaLangSelected
      : this.ariaLangDeselected;
    const filledText = this.ariaDescriptionTranslation
      .replace(/\$position/g, `${this.position}`)
      .replace(/\$selected/g, `${replaceSelected}`);
    return filledText;
  }

  handleIconSize() {
    Array.from(this.hostElement.children).forEach((child) => {
      if (child.tagName.substr(0, 10) === 'SCALE-ICON') {
        child.setAttribute('size', iconSizes[this.size]);
      }
    });
  }

  handleClick = (event: MouseEvent) => {
    event.preventDefault();
    this.selected = !this.selected;
    this.handleIconShape();
    this.scaleClick.emit({ id: this.toggleButtonId, selected: this.selected });
    emitEvent(this, 'scaleClick', {
      id: this.toggleButtonId,
      selected: this.selected,
    });
  };

  handleIconShape = () => {
    if (this.hasScaleIcon) {
      Array.from(this.hostElement.children).forEach((node) => {
        if (node.nodeName.substr(0, 10) === 'SCALE-ICON') {
          if (this.selected) {
            node.setAttribute('selected', 'true');
          } else {
            node.removeAttribute('selected');
          }
        }
      });
    }
  };

  /**
   * Detect whether a child node is a scale icon and contains text.
   * If so, we set `iconPosition` to `after`, if the icon comes after the text.
   */
  setIconPositionProp() {
    const nodes = Array.from(this.hostElement.childNodes).filter((node) => {
      if (node.nodeName.substr(0, 10) === 'SCALE-ICON') {
        this.hasScaleIcon = true;
      }
      // ignore empty text nodes, which are probably due to formatting
      return !(node.nodeType === 3 && node.nodeValue.trim() === '');
    });
    if (
      !this.iconOnly &&
      nodes &&
      nodes.length &&
      nodes[nodes.length - 1] &&
      nodes[nodes.length - 1].nodeName.substr(0, 10) === 'SCALE-ICON'
    ) {
      this.iconPosition = 'after';
    }
  }

  render() {
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}
        <button
          ref={(el) => (this.focusableElement = el)}
          class={this.getCssClassMap()}
          id={this.toggleButtonId}
          onClick={this.handleClick}
          disabled={this.disabled}
          type="button"
          aria-label={this.ariaLabelToggleButton}
          aria-pressed={this.selected}
          part={this.getBasePartMap()}
          aria-description={this.getAriaDescriptionTranslation()}
        >
          <slot />
        </button>
      </Host>
    );
  }

  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }

  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }

  getCssOrBasePartMap(mode: 'basePart' | 'css') {
    const prefix = mode === 'basePart' ? '' : 'toggle-button--';

    return classNames(
      'toggle-button',
      this.size && `${prefix}${this.size}`,
      this.background &&
        `${prefix}${this.background === 'grey' ? 'primary' : 'secondary'}`,
      !this.iconOnly &&
        this.iconPosition &&
        `toggle-button--icon-${this.iconPosition}`,
      this.iconOnly && `${prefix}icon-only`,
      !this.disabled && this.selected && `${prefix}selected`,
      this.radius && `${prefix}${this.radius}`,
      this.colorScheme && `${prefix}${this.colorScheme}`,
      this.variant && `${prefix}${this.variant}`,
      !this.hideBorder && `${prefix}border`
    );
  }
}
