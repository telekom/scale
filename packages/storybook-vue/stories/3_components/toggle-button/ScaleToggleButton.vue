<template>
  <scale-toggle-button
    :size="size"
    :variant="variant"
    :color-scheme="colorScheme"
    :disabled="disabled"
    :selected="selected"
    :icon-only="iconOnly"
    :radius="radius"
    :aria-label="ariaLabel"
    :aria-description-translation="ariaDescriptionTranslation"
    :aria-lang-selected="ariaLangSelected"
    :aria-lang-deselected="ariaLangDeselected"
    :styles="styles"
  >
    <slot></slot>
  </scale-toggle-button>
</template>

<script>
export default {
  props: {
    size: { type: String, default: 'large' },
    variant: { type: String, default: 'primary' },
    colorScheme: { type: String, default: 'magenta' },
    disabled: { type: Boolean, default: false },
    selected: { type: Boolean, default: false },
    iconOnly: { type: Boolean, default: false },
    radius: { type: String },
    ariaLabel: { type: String },
    ariaLangSelected: { type: String, default: 'selected' },
    ariaLangDeselected: { type: String, default: 'deselected' },
    styles: { type: String },
  },
};
</script>
