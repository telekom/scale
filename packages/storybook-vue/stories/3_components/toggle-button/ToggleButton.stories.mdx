import { Meta, Story, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import ScaleToggleButton from './ScaleToggleButton.vue';

<Meta
  title="Beta Components/Toggle Button"
  component={ScaleToggleButton}
  argTypes={{
    variant: {
      control: { type: 'select', options: ['primary', 'secondary', 'ghost'] },
    },
    size: {
      control: { type: 'select', options: ['xs', 'small', 'regular', 'large'] },
    },
    radius: {
      control: { type: 'select', options: ['left', 'right', 'both'] },
    },
	  colorScheme: {
      control: { type: 'select', options: ['light', 'dark'] },
    },
    ariaLangDeselected: {
      control: { type: "text", },
    },
    ariaLangSelected: {
      control: { type: "text", },
    },
  }}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleToggleButton },
  props: {
    label: String,
    icon: String,
    ...ScaleToggleButton.props,
  },
  template: `
  <scale-toggle-button
    :size="size"
    :variant="variant"
    :color-scheme="colorScheme"
    :disabled="disabled"
    :selected="selected"
    :icon-only="iconOnly"
    :radius="radius"
    :aria-label="ariaLabel"
    :aria-lang-selected="ariaLangSelected"
    :aria-lang-deselected="ariaLangDeselected"
    :styles="styles"
  >
    <scale-icon-action-search v-if="icon === 'before'" />
      {{ iconOnly !== true ? label : null }}
      <scale-icon-action-add v-if="iconOnly" />
      <scale-icon-navigation-right v-if="icon === 'after'" />
  </scale-toggle-button>
    `,
});

<div
  style={{
    display: 'inline-flex',
    alignItems: 'center',
    flexDirection: 'column',
    justifyContent: 'space-between',
    width: '100%',
  }}
>
  <h1>Toggle Button</h1>
  <h2>A Toggle Button is meant to be used inside a Toggle Group</h2>
</div>

## Standard

<Canvas withSource="none">
  <Story
    name="Standard"
    args={{
      label: 'Label',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Standard" />

```html
<scale-toggle-button>Label</scale-toggle-button>
```

## Secondary

<Canvas withSource="none">
  <Story
    name="Secondary"
    args={{
      variant: 'secondary',
      label: 'Label',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button variant="secondary">Label</scale-toggle-button>
```

## Disabled

<Canvas withSource="none">
  <Story
    name="Disabled"
    args={{
      disabled: true,
      label: 'Label',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button disabled>Label</scale-toggle-button>
```

## Secondary Disabled

<Canvas withSource="none">
  <Story
    name="Secondary Disabled"
    args={{
      variant: 'secondary',
      disabled: true,
      label: 'Label',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button variant="secondary" disabled>Label</scale-toggle-button>
```

## With Icon Before

<Canvas withSource="none">
  <Story
    name="With Icon Before"
    args={{
      label: 'Label',
      icon: 'before',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button> <scale-icon-action-search /> Label </scale-toggle-button>
```

## With Icon After

<Canvas withSource="none">
  <Story
    name="With Icon After"
    args={{
      label: 'Label',
      icon: 'after',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button>
  Label <scale-icon-navigation-right />
</scale-toggle-button>
```

## Icon Only

<Canvas withSource="none">
  <Story
    name="Icon Only"
    args={{
      iconOnly: true,
      title: 'Search',
      ariaLabel: 'Search',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button>
  <scale-icon-navigation-external-link/>
</scale-toggle-button>
```

## Color Scheme Dark 
### (...click toggle button)

<Canvas withSource="none">
  <Story
    name="Color Scheme Dark"
    args={{
      label: 'Label',
	    colorScheme: 'dark'
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button color-scheme="dark">Label</scale-toggle-button>
```

## Regular With Icon Before

<Canvas withSource="none">
  <Story
    name="Regular With Icon Before"
    args={{
      label: 'Label',
      icon: 'before',
	  size: 'regular'
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button size="regular" >
  <scale-icon-navigation-right /> Label
</scale-toggle-button>
```

## Small With Icon After

<Canvas withSource="none">
  <Story
    name="Small With Icon After"
    args={{
      label: 'Label',
      icon: 'after',
	  size: 'small'
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button size="small" >
  Label <scale-icon-navigation-right />
</scale-toggle-button>
```

## XS

<Canvas withSource="none">
  <Story
    name="XS"
    args={{
      label: 'Label',
	    size: 'xs'
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button size="xs">Label</scale-toggle-button>
```

## Radius Left

<Canvas withSource="none">
  <Story
    name="Radius Left"
    args={{
      label: 'Label',
	    radius: 'left'
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button radius="left">Label</scale-toggle-button>
```

## Radius Both

<Canvas withSource="none">
  <Story
    name="Radius Both"
    args={{
      label: 'Label',
	    radius: 'both'
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button radius="both">Label</scale-toggle-button>
```


## Accessibility

<Canvas withSource="none">
  <Story
    name="Accessibility"
    args={{
      label: 'Label',
      ariaLabel: 'Click me',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button aria-label="Click me">Label</scale-toggle-button>
```


## Accessibility, German

<Canvas withSource="none">
  <Story
    name="Accessibility German"
    args={{
      label: 'Label',
      ariaLabel: 'Drück mich',
      ariaLangSelected: 'selektiert',
      ariaLangDeselected: 'deselektiert',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-toggle-button 
  aria-label="Drück mich"
  aria-lang-selected="selektiert"
  aria-lang-deselected="deselektiert">Label
</scale-toggle-button>
```
