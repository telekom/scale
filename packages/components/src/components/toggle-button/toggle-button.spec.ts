/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { ToggleButton } from './toggle-button';

describe('Toggle Button', () => {
  let page: SpecPage;
  describe('snapshots', () => {
    it('default props', async () => {
      page = await newSpecPage({
        components: [ToggleButton],
        html: `<scale-toggle-button>Test</scale-toggle-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('regular size secondary', async () => {
      page = await newSpecPage({
        components: [ToggleButton],
        html: `<scale-toggle-button size="regular" variant="secondary">Test</scale-toggle-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('black color scheme', async () => {
      page = await newSpecPage({
        components: [ToggleButton],
        html: `<scale-toggle-button color-scheme="black">Test</scale-toggle-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('disabled', async () => {
      page = await newSpecPage({
        components: [ToggleButton],
        html: `<scale-toggle-button size="large" disabled>Test</scale-toggle-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('selected', async () => {
      page = await newSpecPage({
        components: [ToggleButton],
        html: `<scale-toggle-button size="xs" selected>Test</scale-toggle-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('iconOnly', async () => {
      page = await newSpecPage({
        components: [ToggleButton],
        html: `<scale-toggle-button selected radius="both">
			  <scale-icon-action-search></scale-icon-action-search>
			  </scale-toggle-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('iconBefore', async () => {
      page = await newSpecPage({
        components: [ToggleButton],
        html: `<scale-toggle-button radius="right">
				<scale-icon-action-search></scale-icon-action-search>Test
				</scale-toggle-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it('iconAfter', async () => {
      page = await newSpecPage({
        components: [ToggleButton],
        html: `<scale-toggle-button radius="left" styles="color:red">
				  Test<scale-icon-action-search></scale-icon-action-search>
				  </scale-toggle-button>`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });
  describe('classes', () => {
    it('should handle getCssClassMap() and getBasePartMap()', () => {
      const element = new ToggleButton();
      expect(element.getCssClassMap()).toContain('toggle-button');
      expect(element.getCssClassMap()).toContain('toggle-button--secondary');
      expect(element.getCssClassMap()).toContain('toggle-button--icon-before');
      expect(element.getCssClassMap()).toContain('toggle-button--color');

      expect(element.getBasePartMap()).toContain('toggle-button');
      expect(element.getBasePartMap()).toContain('secondary');
      expect(element.getBasePartMap()).toContain('icon-before');
      expect(element.getBasePartMap()).toContain('color');
      element.iconOnly = true;
      expect(element.getCssClassMap()).toContain('toggle-button--icon-only');
      expect(element.getBasePartMap()).toContain('icon-only');
    });
  });

  describe('functions', () => {
    beforeEach(async () => {
      page = await newSpecPage({
        components: [ToggleButton],
        html: `<scale-toggle-button>
						Test<scale-icon-action-search></scale-icon-action-search>
					</scale-toggle-button>`,
      });
    });
    it('setIconPositionProp() returns "after"', async () => {
      page.rootInstance.setIconPositionProp();
      await page.waitForChanges();
      expect(page.rootInstance.iconPosition).toBe('after');
    });
    it('"selected" Prop toggles on click', async () => {
      page.rootInstance.selected = false;
      const toggleButton = page.root.shadowRoot.querySelector('.toggle-button');
      await page.waitForChanges();
      toggleButton.dispatchEvent(new Event('click'));
      await page.waitForChanges();
      expect(page.rootInstance.selected).toBe(true);
      toggleButton.dispatchEvent(new Event('click'));
      await page.waitForChanges();
      expect(page.rootInstance.selected).toBe(false);
    });
  });
});
