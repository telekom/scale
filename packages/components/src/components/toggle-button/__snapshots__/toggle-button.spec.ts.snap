// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Button snapshots black color scheme 1`] = `
<scale-toggle-button color-scheme="black" icon-position="before" toggle-button-id="toggle-button-2">
  <mock:shadow-root>
    <button aria-description="deselected" class="toggle-button toggle-button--black toggle-button--border toggle-button--color toggle-button--icon-before toggle-button--regular toggle-button--secondary" id="toggle-button-2" part="toggle-button regular secondary toggle-button--icon-before black color border" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Test
</scale-toggle-button>
`;

exports[`Toggle Button snapshots default props 1`] = `
<scale-toggle-button icon-position="before" toggle-button-id="toggle-button-0">
  <mock:shadow-root>
    <button aria-description="deselected" class="toggle-button toggle-button--border toggle-button--color toggle-button--icon-before toggle-button--regular toggle-button--secondary" id="toggle-button-0" part="toggle-button regular secondary toggle-button--icon-before color color border" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Test
</scale-toggle-button>
`;

exports[`Toggle Button snapshots disabled 1`] = `
<scale-toggle-button disabled="" icon-position="before" size="large" toggle-button-id="toggle-button-3">
  <mock:shadow-root>
    <button aria-description="deselected" class="toggle-button toggle-button--border toggle-button--color toggle-button--icon-before toggle-button--large toggle-button--secondary" disabled="" id="toggle-button-3" part="toggle-button large secondary toggle-button--icon-before color color border" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Test
</scale-toggle-button>
`;

exports[`Toggle Button snapshots iconAfter 1`] = `
<scale-toggle-button icon-position="after" radius="left" styles="color:red" toggle-button-id="toggle-button-7">
  <mock:shadow-root>
    <style>
      color:red
    </style>
    <button aria-description="deselected" class="toggle-button toggle-button--border toggle-button--color toggle-button--icon-after toggle-button--left toggle-button--regular toggle-button--secondary" id="toggle-button-7" part="toggle-button regular secondary toggle-button--icon-after left color color border" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Test
  <scale-icon-action-search size="22"></scale-icon-action-search>
</scale-toggle-button>
`;

exports[`Toggle Button snapshots iconBefore 1`] = `
<scale-toggle-button icon-position="before" radius="right" toggle-button-id="toggle-button-6">
  <mock:shadow-root>
    <button aria-description="deselected" class="toggle-button toggle-button--border toggle-button--color toggle-button--icon-before toggle-button--regular toggle-button--right toggle-button--secondary" id="toggle-button-6" part="toggle-button regular secondary toggle-button--icon-before right color color border" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  <scale-icon-action-search size="22"></scale-icon-action-search>
  Test
</scale-toggle-button>
`;

exports[`Toggle Button snapshots iconOnly 1`] = `
<scale-toggle-button icon-position="after" radius="both" selected="" toggle-button-id="toggle-button-5">
  <mock:shadow-root>
    <button aria-description="selected" aria-pressed="" class="toggle-button toggle-button--border toggle-button--both toggle-button--color toggle-button--icon-after toggle-button--regular toggle-button--secondary toggle-button--selected" id="toggle-button-5" part="toggle-button regular secondary toggle-button--icon-after selected both color color border" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  <scale-icon-action-search selected="true" size="22"></scale-icon-action-search>
</scale-toggle-button>
`;

exports[`Toggle Button snapshots regular size secondary 1`] = `
<scale-toggle-button icon-position="before" size="regular" toggle-button-id="toggle-button-1" variant="secondary">
  <mock:shadow-root>
    <button aria-description="deselected" class="toggle-button toggle-button--border toggle-button--color toggle-button--icon-before toggle-button--regular toggle-button--secondary" id="toggle-button-1" part="toggle-button regular secondary toggle-button--icon-before color secondary border" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Test
</scale-toggle-button>
`;

exports[`Toggle Button snapshots selected 1`] = `
<scale-toggle-button icon-position="before" selected="" size="xs" toggle-button-id="toggle-button-4">
  <mock:shadow-root>
    <button aria-description="selected" aria-pressed="" class="toggle-button toggle-button--border toggle-button--color toggle-button--icon-before toggle-button--secondary toggle-button--selected toggle-button--xs" id="toggle-button-4" part="toggle-button xs secondary toggle-button--icon-before selected color color border" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Test
</scale-toggle-button>
`;
